lib/appRelationsDiscovery.js

```diff
@@ -0,0 +1,65 @@
+const {validate} = require('./utils');
+
+const supportedSyntax = ['default','mermaid'];
+
+// eslint-disable-next-line sonarjs/cognitive-complexity
+async function getRelations(asyncApiDocs,requestedSyntax = 'default') {
+  if (typeof asyncApiDocs !== 'object') throw new Error('You must pass an array of AsyncApiDocuments on which you wish to discover the relations between');
+  if (!supportedSyntax.includes(requestedSyntax)) return;
+
+  const parsedAsyncApiDocs = await validate(asyncApiDocs);
+  const metrics = new Map();
+  
+  parsedAsyncApiDocs.forEach(doc => {
+    if (doc.hasServers()) {
+      const servers = doc.servers();
+
+      for (const [, credentials] of Object.entries(servers)) {
```

Comment:
If you need only values you can use `.values()` function 

```suggestion
      for (const credentials of Object.values(servers)) {
```

Reply:
I had it like `for (const [serverName, credentials] of Object.entries(servers)) ` and was thinking to add `serverName` in         ``const slug = `${credentials.url()},${credentials.protocol()}`` so as to make it more unique. Like what if two servers named 'production' and 'development' have the same url and protocol, in that case, slug would remain same and services would double up in the Map.

Reply:
But `serverName` is only a name for serve, it's something like metadata for spec, it can be named in different way in other spec, so as you wrote you can end with double items in map for this same server. I think that we should operate only on url. From you comment I also find some problems, because someone can describe server's url with parameters like `server/{parameter}` and define for this parameter some enum values. Not now, but in the next weeks (maybe at the end of gsoc) we should check if this server is exactly this same as in another spec. I hope that you understand :)

Reply:
> I think that we should operate only on url and protocol
> someone can describe server's url with parameters like `server/{parameter}` and define for this parameter some enum values.

Yeah! you're right. I just checked out the specs and saw that we could have it like this(`server/{parameter}`) too.

```
servers:
- url: '{username}.gigantic-server.com:{port}/{basePath}'
  description: The production API server
  protocol: secure-mqtt
  variables:
    username:
      # note! no enum here means it is an open value
      default: demo
      description: This value is assigned by the service provider, in this example `gigantic-server.com`
    port:
      enum:
        - '8883'
        - '8884'
      default: '8883'
    basePath:
      # open meaning there is the opportunity to use special base paths as assigned by the provider, default is `v2`
      default: v2
```
This might be a big concern coz like here, we have to configure the flow to support this case too, also need to iterate with multiple `ports: 8883, 8884`. I will keep this in mind, and would definitely fix it in the next weeks. For now, I think we can move forward. What do you suggest? 🤔 

Shall I create an issue now only regarding this?

Reply:
Yep, you should create issue for that to not forget about it :)

Reply:
Duly noted, will keep this in mind.

> But `serverName` is only a name for server, it's something like metadata for server, it can be named in different way in other spec, so as you wrote you can end with double items in map for this same server

Yeah, `serverName` could be anything but won't it be a bug if we end up doubling up items for the same server.?

Also, normally user won't have same protocol and url for different defined servers but I was just considering every possible case.

Reply:
> But `serverName` is only a name for server, it's something like metadata for server, it can be named in different way in other spec, so as you wrote you can end with double items in map for this same server

Yeah, `serverName` could be anything but won't it be a bug if we end up doubling up items for the same server?

Also, normally user won't have same protocol and url for different defined servers within the same file but I was just considering every possible case.

Reply:
> Also, normally user won't have same protocol and url for different defined servers within the same file but I was just considering every possible case.

I don't think so as you. Different servers should be defined one time in spec.

> Yeah, serverName could be anything but won't it be a bug if we end up doubling up items for the same server?

It's true that it won't be a "normal" bug but for me it will be a "logical" bug. You operate on the servers of the given spec. In one spec you can describe one server with this same url and protocol but with (eg) `server1` name, in the second you can describe this same server with url and protocol but with `server2` name. When you will use `serverName` for the `relation` key you can end up with doubled values in the relations Map, with different pub/sub items. 

```yaml
# one spec
servers
  server1:
    url: someUrl
    protocol: kafka
channels:
  someChannel: ...

# second spec
servers:
  server2:
    url: someUrl
    protocol: kafka
channels:
  someAnotherChannel: ...
``` 

When you will use the `serverName` you will have two relations, for `server1` and `server2` and `server1` will have only channels related to spec with `server1`, and `someAnotherChannel` channel will be related only to `server2` when this is the same server :)

Reply:
You can create issue for that and we can discuss about it in the new issue :)

Reply:
I think you misunderstood me here. What you described is totally clear to me.

But, I was talking about a case like this one.

```yaml
#one spec
  servers: 
    production:
        url: mqtt://localhost:1883
        protocol: mqtt
    development:
        url: mqtt://localhost:1883
        protocol: mqtt
  channels:
  someAnotherChannel: ...
```
If this is the case then we would iterate 2 times(for production and development) and then double up the applications described for this AsyncAPI document.

Reply:
Ahh sorry... I think about `serverName` again and again...  Of course, you're right, in current logic you will have doubled pub/sub in your app, but I don't think so that it should be hard to fix, but not now :) You can create issue for that and handle it in the next PRs.

Reply:
Yeah, it could be a potential bug coz user should never have define the spec having two servers with same url and protocol in the first place. Would open an issue for the same 👍🏽